src/stash/mod.rs

```diff
@@ -25,51 +28,54 @@ impl<'a, I> Drop for Extend<'a, I>
     }
 }
 
-impl<'a, I> Iterator for Extend<'a, I>
+impl<'a, I, Ix: Index> Iterator for Extend<'a, I, Ix>
     where I: Iterator,
           I::Item: 'a
 {
-    type Item = usize;
+    type Item = Ix;
 
-    fn next(&mut self) -> Option<usize> {
+    fn next(&mut self) -> Option<Ix> {
         self.iter.next().map(|v| self.stash.put(v))
     }
     fn size_hint(&self) -> (usize, Option<usize>) {
         self.iter.size_hint()
     }
 }
 
-impl<'a, I> ExactSizeIterator for Extend<'a, I>
+impl<'a, I, Ix: Index> ExactSizeIterator for Extend<'a, I, Ix>
     where I: ExactSizeIterator,
           I::Item: 'a
 {
 }
 
-impl<'a, I> DoubleEndedIterator for Extend<'a, I>
+impl<'a, I, Ix: Index> DoubleEndedIterator for Extend<'a, I, Ix>
     where I: DoubleEndedIterator,
           I::Item: 'a
 {
-    fn next_back(&mut self) -> Option<usize> {
+    fn next_back(&mut self) -> Option<Ix> {
         self.iter.next_back().map(|v| self.stash.put(v))
     }
 }
 
 /// Iterator over the `(index, &value)` pairs.
-pub struct Iter<'a, V: 'a> {
+pub struct Iter<'a, V: 'a, Ix: Index> {
     inner: iter::Enumerate<slice::Iter<'a, Entry<V>>>,
     len: usize,
+    _marker: marker::PhantomData<Ix>,
```

Comment:
Technically, this should be `marker::PhantomData<fn() -> Ix>` to indicate that we don't actually store any `Ix`s although I'm not sure if it makes any difference...